lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -1969,6 +1969,34 @@ static SILValue getTensorProtocolHandleMember(SILValue v, SILLocation loc,
   return v;
 }
 
+/// Given a type, recursively collect all innermost element types if it's an
+/// aggregate of TensorHandle's or dtypes.
+static bool collectInnermostTensorFlowDTypes(
+    CanType type, SmallVectorImpl<SymbolicValue> &result) {
+  if (isTensorFlowDType(type)) {
+    result.push_back(SymbolicValue::getMetatype(type));
+  } else if (tf::isTensorHandle(type)) {
+    auto eltType = getTensorHandleElementType(type)->getCanonicalType();
+    assert(tf::isTensorFlowDType(eltType));
+    result.push_back(SymbolicValue::getMetatype(eltType));
+  } else if (auto tupleTy = type->getAs<TupleType>()) {
+    for (auto eltTy : tupleTy->getElementTypes())
+      if (!collectInnermostTensorFlowDTypes(eltTy->getCanonicalType(), result))
+        return false;
+  } else if (auto *decl = type->getStructOrBoundGenericStruct()) {
+    for (auto *member : decl->getStoredProperties()) {
+      auto subMap =
+          type->getMemberSubstitutionMap(member->getModuleContext(), member);
+      auto memberTy = member->getType().subst(subMap)->getCanonicalType();
+      if (!collectInnermostTensorFlowDTypes(memberTy, result))
+        return false;
+    }
+  } else {
+    return false;
+  }
+  return true;
```

Comment:
These three lines are confusing.  I'd suggest putting the "return true" at the end of each "if branch" above, turning the "if else" chain into a bunch of unchained "ifs".  That will make it more clear that each case is an independent case and that reaching the end of the function is failure.